src/framebounds.c

```diff
@@ -34,55 +54,20 @@ void adjust_frame_boundaries (const struct plum_image * image, struct plum_recta
-        bool adjust = true;                                                                                                                                 \
-        for (size_t remaining = (size_t) boundaries[frame].top * image -> width; remaining; remaining --)                                                   \
-          if (image -> data ## bits[index ++] != empty_color) goto done ## bits;                                                                            \
-        if (boundaries[frame].left || boundaries[frame].width != image -> width)                                                                            \
-          for (uint_fast32_t row = 0; row < boundaries[frame].height; row ++) {                                                                             \
-            for (uint_fast32_t col = 0; col < boundaries[frame].left; col ++) if (image -> data ## bits[index ++] != empty_color) goto done ## bits;        \
-            index += boundaries[frame].width;                                                                                                               \
-            for (uint_fast32_t col = boundaries[frame].left + boundaries[frame].width; col < image -> width; col ++)                                        \
-              if (image -> data ## bits[index ++] != empty_color) goto done ## bits;                                                                        \
-          }                                                                                                                                                 \
-        else                                                                                                                                                \
-          index += (size_t) boundaries[frame].height * image -> width;                                                                                      \
-        for (size_t remaining = (size_t) (image -> height - boundaries[frame].top - boundaries[frame].height) * image -> width; remaining; remaining --)    \
-          if (image -> data ## bits[index ++] != empty_color) goto done ## bits;                                                                            \
-        adjust = false;                                                                                                                                     \
-        done ## bits:                                                                                                                                       \
-        if (adjust) boundaries[frame] = (struct plum_rectangle) {.left = 0, .top = 0, .width = image -> width, .height = image -> height};                  \
-      }                                                                                                                                                     \
-    while (false)
+    #undef notempty
+    #define notempty(bits) (image -> data ## bits[index ++] != empty_color)
```

Comment:
rgbds has already been converted to C++; next up, libplum, with templates and anonymous closures replacing all these macros. >:3